cross-project `ref` + model `access` + `dependencies.yml` (#3577)
dbt-labs/docs.getdbt.com: website/docs/docs/collaborate/govern/model-access.md

As we continue to develop multi-project collaboration, `access: public` will mean that other teams are allowed to start taking a dependency on that model. This assumes that they've requested, and you've granted them access, to select from the underlying dataset.

### How do I ref a model from another project?

<VersionBlock lastVersion="1.5">

In dbt Core v1.5, the only way to reference a model from another project is by installing that project as a package. There is no way to enforce access distinctions between `protected` and `public` models.

> This changes in v1.6. Select it from the dropdown above.

Reviewer: ```suggestion
This changes in v1.6. Select v1.6 from the version dropdown to view more information. 
```

</VersionBlock>

<VersionBlock firstVersion="1.6">

It is possible to `ref` a model from another project in two ways:
1. As a "project" dependency, via "cross-project `ref`" (a feature of dbt Cloud Enterprise)

> 2. As a "package" dependency, whereby all source code from that project is installed into your own

Reviewer: Looks like we just need to finish this thought - own project?

Author: I definitely meant to write `your own project` here — although we're using `project` to mean so many different things. It's almost more like:
- your own working directory
- your own environment
- your own dbt runtime
- ...

Reviewer: I like `environment`!

> 2. As a "package" dependency, whereby all source code from that project is installed into your own

Reviewer: ```suggestion
2. As a "package" dependency, whereby all source code from that project is installed into your own environment
```

See ["Project Dependencies"](project-dependencies) for an explanation of the advantages of each approach.

#### How do I restrict access to models defined in a package?

> The maintainer of a project that can be installed as a package may choose whether to restrict other models' `ref` access to only its public models. The package maintainer does this by setting a `restrict_access` config to `True`. By default, the config is `False`, meaning that:

Reviewer: ```suggestion
The maintainer of a project that can be installed as a package may choose whether to restrict other models' `ref` access to only its public models. The package maintainer sets a `restrict_access` config to `True`. By default, the config is `False`, meaning that:
```

> - Models in the package with `access: protected` may be referenced by models in the root project, as if they were defined in the same project

Reviewer: ```suggestion
- Models in the package with `access: protected` may be referenced by models in the root project as if they were defined in the same project
```

- Models in the package with `access: private` may be referenced by models in the root project, so long as they also have the same `group` config

> When `restrict_access: True`, any `ref` from outside the package to a protected or private model in that package will fail. Only models with `access: public` may be accessed outside the package.

Reviewer 2: This was implemented as [restrict-access](https://github.com/dbt-labs/dbt-core/pull/7962/files#diff-2c495a0925e5096258f6e87aa1784f8139a23c940601d816eaa8db5e5ad9975aR176) (not restricted_access) for naming consistency with other dbt_project.yml-level configs.

Author: Ah, yep, makes sense. It would be nice if we were globally consistent about snake versus kebab casing (snake should really win)

> When `restrict_access: True`, any `ref` from outside the package to a protected or private model in that package will fail. Only models with `access: public` may be accessed outside the package.

Reviewer: ```suggestion
When `restrict-access: True`, any `ref` from outside the package to a protected or private model in that package will fail. Only models with `access: public` may be accessed outside the package.
```

As we continue to develop multi-project collaboration, `access: public` will mean that other teams are allowed to start taking a dependency on that model. This assumes that they've requested, and you've granted them access, to select from the underlying dataset.

### How do I ref a model from another project?

<VersionBlock lastVersion="1.5">

In dbt Core v1.5 (and earlier versions), the only way to reference a model from another project is by installing that project as a package, including its full source code. It is not possible to restrict references across projects based on model `access`.

For more control over per-model access across projects, select v1.6 (or newer) from the version dropdown.

</VersionBlock>

<VersionBlock firstVersion="1.6">

> It is possible to `ref` a model from another project in two ways:

Reviewer 3: ```suggestion
You can `ref` a model from another project in two ways:
```

> 1. As a "project" dependency, a feature of dbt Cloud Enterprise. The reference is resolved via a metadata service, and it enables effective collaboration across teams and at scale. Read more about ["Project Dependencies"](project-dependencies).

Reviewer 3: ```suggestion
1. [Project dependency](/docs/collaborate/govern/project-dependencies) &mdash; In dbt Cloud Enterprise, you can use project dependencies to `ref`  a model. This method utilizes a metadata service to resolve the reference and allows efficient collaboration across teams and at scale.
```

> 2. As a "package" dependency. This requires installing the other project as a package, including its full source code, as well as its upstream dependencies.

Reviewer 3: ```suggestion
2. Package dependency &mdash;  Another way to `ref` a model from another project is to treat the other project as a package dependency. This requires installing the other project as a package, including its full source code, as well as its upstream dependencies.

```

Author: nice! I like this
